### ISSUE TITLE
Use Frame Timing API

### ISSUE THREAD

**Issue by Richienb** [Richienb/refresh-rate#1]
When supported, use the [Frame Timing API](https://developer.mozilla.org/en-US/docs/Web/API/Frame_Timing_API/Using_the_Frame_Timing_API) to listen for `frame` events.

**Comment by mirrorlink** [comment]
hey @Richienb i just wanted to thank you for this package, and ask for you knowledge.

I've developed a game called Wirewalk() with Electron. ( https://store.steampowered.com/app/1636700/Wirewalk/ )

But the problem is that the Box2D physics depends on rAF (GPU sync) to sync and step properly.

From my findings, it seems that when there is an error on measurement, it is most likely to less fps, right? I think it will very rarely say that the FPS is much higher

So, I'm using this algorithm, see if you think I should tweak it somehow:

```
        measureRefreshRate({sampleCount: 400}).then((rate: number) =>
        {
            console.log('refresh rate', rate);

            if(rate < 30) //this does not look right.. lets try that again
            {
                console.log('too low fps, try again');
                this.measureFPS();
            }
            else if(rate <= 64) //60hz
            {
                console.log('60fps');
                //already default
            }
            else if(rate <= 79) //75hz
            {
                console.log('75fps');
                cc.game.setFrameRate(75);
            }
            else if( rate <= 124) //weird-monitor limbo, sync on CPU at 60fps
            {
                //@ts-ignore
                window.useCPUSync = true;

                console.log('weird-monitor limbo');
                cc.game.setFrameRate(60);
            }
            else if( rate <= 148) //144hz
            {
                console.log('144fps');
                cc.game.setFrameRate(144);
            }
            else if( rate < 169) //165hz
            {
                console.log('165fps');
                cc.game.setFrameRate(165);
            }
            else //too high fps, cap on CPU at 120fps
            {
                //@ts-ignore
                window.useCPUSync = true;

                console.log('too high fps');
                cc.game.setFrameRate(120);
            }
        });
```

Thank you!

<blockquote><img src="https://cdn.cloudflare.steamstatic.com/steam/apps/1636700/capsule_616x353.jpg?t=1632395318" width="48" align="right"><div><strong><a href="https://store.steampowered.com/app/1636700/Wirewalk/">Wirewalk()↳ on Steam</a></strong></div><div>Wirewalk()↳ is an adventure where you raid virus-riddled systems for the lols.</div></blockquote>

**Comment by Richienb** [comment]
Instead of setting fps based on fps, the best strategy is to allow for limitless fps and interpolate movement based on the time between frames.

**Comment by mirrorlink** [comment]
> Instead of setting fps based on fps, the best strategy is to allow for limitless fps and interpolate movement based on the time between frames.

This works for animation, but not for physics. You can't 'catch up' on physics, otherwise you can overshoot walls if a frame takes too long, or change behavior on each frame because of a different step. it has to be a fixed step.

One solution is to use an accumulator and fixed-step, but since my game is pixel-perfect, not running physics on some frames / running different number of steps on some frames causes it to flick very badly.

Thats why I'm founding this to be the only solution

**Comment by Richienb** [comment]
Good point. You might be able to constantly measure fps using `requestAnimationFrame` and measure how many times it is called in a second every second.

**Comment by mirrorlink** [comment]
I see. I that's a good Idea. I think I will be constantly measuring it, and I will take the results and keep a counter on each of the cases.  Then change framerate if one counter surpasses another.  I may be able to ignore measurement too if the screen is minimized, unless its the first one.
